src/corelibs/U2Core/src/util/MsaRowUtils.cpp

```diff
@@ -432,4 +432,80 @@ void MsaRowUtils::removeTrailingGapsFromModel(qint64 length, QVector<U2MsaGap> &
+        SAFE_POINT(isValidGapModel, "Invalid gap model", {});
+        if (gap == nullptr || gap->startPos > pos) {
+            // Processing core sequence part. At this point no gaps left or the next gap starts after 'pos'.
+            int corePartLength = gap == nullptr ? coreLength - corePos : gap->startPos - pos;
+            SAFE_POINT(corePos + corePartLength <= coreLength, "Invalid position in core sequence!", {});
+            if (region.startPos < pos + corePartLength) {  // Add core sub-sequence to the result.
+                int nChars;
+                int corePartOffset;
+                if (region.startPos <= pos) {  // Add a prefix or whole core sequence region.
+                    nChars = qMin(corePartLength, regionEnd - pos);
+                    corePartOffset = 0;
+                } else if (regionEnd > pos + corePartLength) {  // Add suffix of the core sequence region.
+                    nChars = pos + corePartLength - region.startPos;
+                    corePartOffset = corePartLength - nChars;
+                } else {  // Add middle part of the core sequence region.
+                    SAFE_POINT(region.startPos > pos && regionEnd <= pos + corePartLength, "Invalid middle region coordinates", {};)
+                    nChars = region.length;
+                    corePartOffset = region.startPos - pos;
+                }
+                SAFE_POINT(nChars > 0 && corePartOffset >= 0 && corePartOffset + nChars <= corePartLength && result.length() + nChars <= region.length,
+                           "Invalid sub-region in the core region",
+                           {});
+                result.append(coreSequence + corePos + corePartOffset, nChars);
+            }
+            corePos += corePartLength;
+            pos += corePartLength;
+            SAFE_POINT(corePos <= coreLength, "Core sequence overflow!", {});
+        } else {
+            // Processing gap region.
+            int gapPartLength = (int)gap->length;
+            if (region.startPos < pos + gapPartLength) {
+                int nGaps;
+                if (region.startPos <= pos) {  // Add a prefix or whole gap region.
+                    nGaps = qMin(gapPartLength, regionEnd - pos);
+                } else if (regionEnd > pos + gapPartLength) {  // Add suffix of the gap region
+                    nGaps = pos + gapPartLength - region.startPos;
+                } else {  // Add middle part of the gap region.
+                    nGaps = region.length;
+                }
+                SAFE_POINT(nGaps <= gapPartLength && result.length() + nGaps <= region.length, "Invalid gap region", {});
+                result.append(nGaps, U2Msa::GAP_CHAR);
+            }
+            pos += gap->length;
+            gapIndex++;
+        }
+        SAFE_POINT(pos < region.startPos ? result.isEmpty() : result.length() <= pos - region.startPos, "Invalid result region after processing gap.", {});
```

Comment:
Sometimes you try to fill the only line with a number or expressions and it makes them difficult to read. There is no need of doing that, we do not have lines limitations! I think, it's better either to separate this stuff on several `SAFE_POINT`s or use the usual, very base and common, but so much pretty understandable `if-else`

Reply:
You right, but I do it only for SAFE_POINTS. SAFE_POINTS are not checks (which is a part of "normal" logic) so I try to keeping them as short as possible, to give priority to the "real" code.